### PR TITLE
OBPIH-6497 Fix comparing invoice quantities in UoM and in standard Uom

### PR DESCRIPTION
**Pull request by awalkowiak**
### :sparkles: Description of Change

**Link to GitHub issue or Jira ticket:**
https://pihemr.atlassian.net/browse/OBPIH-6497

**Description:**
This is a fix to issues found during QA. Turned out that we had a few places that were mixing up quantities in UoM and quantities in standard UoM. 

---
### :chart_with_upwards_trend: Test Plan

> *We require that all code changes come paired with a method of testing them. Please select which of the following testing approaches you've included with this change:*

- [ ] Frontend automation tests (unit)
- [X] Backend automation tests ([unit](https://pihemr.atlassian.net/wiki/spaces/OB/pages/3013869579/Backend+Unit+Testing+Standards), [API](https://pihemr.atlassian.net/wiki/spaces/OB/pages/3013738522/Backend+Integration+Testing+Standards), smoke)
- [ ] End-to-end tests ([Playwright](https://pihemr.atlassian.net/wiki/spaces/OB/pages/2942828546/Knowledge+Transfer+on+automated+testing+with+Playwright))
- [x] Manual tests (please describe below)
- [ ] Not Applicable

**Description of test plan (if applicable):**

---
### :white_check_mark: Quality Checks

> *Please confirm and check each of the following to ensure that your change conforms to the coding standards of the project:*

- [X] The pull request title is prefixed with the issue/ticket number (Ex: `[OBS-123]` for Jira, `[#0000]` for GitHub, or `[OBS-123, OBPIH-123]` if there are multiple), or with `[N/A]` if not applicable
- [X] The pull request description has enough information for someone without context to be able to understand the change and why it is needed
- [X] The change has tests that prove the issue is fixed / the feature works as intended

### CODE REVIEW

**Review by kchelstowski — APPROVED** [review]

**Comment by EWaterman** [comment]
@awalkowiak can you explain the difference between standard UOM vs. non-standard and when we want to use each? I'm not clear which is supposed to be used in any given scenario

File: grails-app/domain/org/pih/warehouse/order/OrderAdjustment.groovy [hunk 1]
```diff
@@ -108,8 +108,23 @@ class OrderAdjustment implements Serializable, Comparable<OrderAdjustment> {
...
 
+    /**
+     * Adjustment can be on a regular invoice if it is not canceled or if it is canceled, but has prepayment
+     * */
+    Boolean canBeOnRegularInvoice() {
+        if (canceled) {
+            return hasPrepaymentInvoice
+        }
+
+        return true
```
Comment (EWaterman) [inline grails-app/domain/org/pih/warehouse/order/OrderAdjustment.groovy:119]: can be simplified to `return !canceled || hasPrepaymentInvoice` if you want

File: grails-app/domain/org/pih/warehouse/order/OrderItem.groovy [hunk 2]
```diff
@@ -320,16 +321,29 @@ class OrderItem implements Serializable, Comparable<OrderItem> {
...
 
+    /**
+     * Item can be on a regular invoice if it is not canceled or if it is canceled, but has prepayment
+     * */
+    Boolean canBeOnRegularInvoice() {
+        if (canceled) {
+            return hasPrepaymentInvoice
+        }
+
+        return true
```
Comment (EWaterman) [inline grails-app/domain/org/pih/warehouse/order/OrderItem.groovy:332]: same here. Can be simplified to return !canceled || hasPrepaymentInvoice if you want

**Comment by awalkowiak** [comment]
@EWaterman sure: 
Order items have quantity, unit of measure, and quantity per unit of measure that is not always equal to one. Things are ordered and invoiced in "unit of measure" quantity (which is equal to `quantity`), but shipments and receipts are done in standard uom quantity (which is equal to `quantity from order * quantity per uom`), so if we are comparing invoiced quantity with the shipped quantity we have to be sure we are comparing correct values (in this case either both compared quantities have to be `quantity in standard uom` or both to be in `uom quantity`). 

I know that there are some plans for the upcoming release(s) to make this less confusing.

File: grails-app/domain/org/pih/warehouse/order/Order.groovy [hunk 3]
```diff
@@ -495,13 +495,11 @@ class Order implements Serializable {
     }
 
     Boolean hasInvoiceableOrderItem() {
-        return orderItems.any { it.invoiceable }
+        return orderItems.any { it.canBeOnRegularInvoice() && it.invoiceable }
```
Comment (EWaterman) [inline grails-app/domain/org/pih/warehouse/order/Order.groovy:498]: is there a reason to not move the `canBeOnRegularInvoice` methods to within the `invoiceable` methods? Do we ever need to know if an item/adjustment is invoiceable outside of the context of generating a regular invoice for it?

Comment (awalkowiak) [reply grails-app/domain/org/pih/warehouse/order/Order.groovy:498]: It has to be separate for checking if the order `isFullyInvoiceble`. The `canBeOnRegularInvoice` still could be part of `invoiceable` for this case that you pointed out, but since I needed it separately for another place, I decided not to combine these together (at least for now).

File: grails-app/domain/org/pih/warehouse/order/OrderAdjustment.groovy (hunk 1 above)
Comment (awalkowiak) [reply grails-app/domain/org/pih/warehouse/order/OrderAdjustment.groovy:119]: Yeah, when I was doing the improvements to fix some edge cases, I wanted to keep it as clear as possible, to avoid any further confusion. I want to add a separate subtask for adding more tests (and improving currently existing ones) that will be done after the solution is working properly, then this could be changed to a one-liner there, if this is preferable.
